pkg/storage/utils/decomposedfs/tree/propagator/async.go

```diff
@@ -0,0 +1,427 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package propagator
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cs3org/reva/v2/pkg/appctx"
+	"github.com/cs3org/reva/v2/pkg/logger"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/lookup"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/metadata"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/metadata/prefixes"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/node"
+	"github.com/google/renameio"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/rogpeppe/go-internal/lockedfile"
+	"github.com/rs/zerolog"
+	"github.com/shamaton/msgpack/v2"
+)
+
+const propagationDelay = 5 * time.Second
+
+type AsyncPropagator struct {
+	treeSizeAccounting bool
+	treeTimeAccounting bool
+	lookup             lookup.PathLookup
+}
+
+type Change struct {
+	SyncTime time.Time
+	SizeDiff int64
+}
+
+func NewAsyncPropagator(treeSizeAccounting, treeTimeAccounting bool, lookup lookup.PathLookup) AsyncPropagator {
+	p := AsyncPropagator{
+		treeSizeAccounting: treeSizeAccounting,
+		treeTimeAccounting: treeTimeAccounting,
+		lookup:             lookup,
+	}
+
+	log := logger.New()
+
+	log.Info().Msg("async propagator starting up...")
+
+	// spawn a goroutine that watches for stale .processing dirs and fixes them
+	go func() {
+		if !p.treeTimeAccounting && !p.treeSizeAccounting {
+			// no propagation enabled
+			log.Debug().Msg("propagation disabled or nothing to propagate")
+			return
+		}
+
+		changesDirPath := filepath.Join(p.lookup.InternalRoot(), "changes")
+		doSleep := false // switch to not sleep on the first iteration
+		for {
+			if doSleep {
+				time.Sleep(5 * time.Minute)
+			}
+			doSleep = true
+			log.Debug().Msg("scanning for stale .processing dirs")
+
+			entries, err := filepath.Glob(changesDirPath + "/**/*")
+			if err != nil {
+				continue
+			}
+
+			for _, e := range entries {
+				changesDirPath := e
+				entry, err := os.Stat(changesDirPath)
+				if err != nil {
+					continue
+				}
+				// recover all dirs that haven't been propagated within 3 minutes
+				if !entry.IsDir() || time.Now().Before(entry.ModTime().Add(3*time.Minute)) {
+					continue
+				}
+
+				go func() {
+					if !strings.HasSuffix(changesDirPath, ".processing") {
+						// first rename the existing node dir
+						err = os.Rename(changesDirPath, changesDirPath+".processing")
+						if err != nil {
+							return
+						}
+						changesDirPath += ".processing"
+					}
+
+					log.Debug().Str("dir", changesDirPath).Msg("propagating stale .processing dir")
+					parts := strings.SplitN(entry.Name(), ":", 2)
+					now := time.Now()
+					_ = os.Chtimes(changesDirPath, now, now)
+					p.propagate(context.Background(), parts[0], strings.TrimSuffix(parts[1], ".processing"), true, *log)
```

Comment:
We need to check for `len(parts)` first. Otherwise this can panic under strange circumstances